2_process/src/match_sites_reaches.R

```diff
@@ -0,0 +1,65 @@
+#' original author: David Watkins
+#' from https://code.usgs.gov/wwatkins/national-site-reach/-/blob/master/R/match_sites_reaches.R
+#' modified by: Jeff Sadler
+#' Match each site with a reach (seg_id/COMID)
+get_site_flowlines <- function(reach_sf, sites, sites_crs, max_matches = 1, search_radius = 0.1) {
+  #' 
+  #' @description Function to match reaches (flowlines) to point sites 
+  #'
+  #' @param reach_sf sf object of reach polylines with column "subsegid" and in WGS84
+  #' @param sites dataframe with columns "lat" "lon"
+  #' @param sites_crs the crs os the sites table (i.e., 4269 for NAD83)
+  #' @param max_matches the maximum number of segments that a point can match to
+  #' @param search_radius the maximum radius in same units as sf object
+  #' within which a segment will match (segments outside of the radius will not match)
+  #'
+  #' @value A data frame the same columns as the sites input dataframe with additional columns
+  #' of "segidnat", "subsegid" and "offset" where "offset" is the distance (in degrees) between the point
+  #' and matching flowline
+  
+  # set up NHDPlus fields used by get_flowline_index
+  # Note: that we are renaming the subsegid column to COMID because the nhdplusTools
+  # function requires an sf object with a "COMID" column. This does not have anything
+  # to do with the actual nhd COMIDs 
+  # Note: the `ToMeas` and `FromMeas` are also required columns for the nhdplusTools 
+  # function. Since we are using our own reaches and not the nhd, these do not have 
+  # the same meaning as they would if we were using the nhd
+  reaches_nhd_fields <- reach_sf %>%
+    rename(COMID = subsegid) %>%
+    mutate(REACHCODE = COMID, ToMeas = 100, FromMeas = 100) %>%
+    st_as_sf()
+  
+  sites_sf <- sites %>% rowwise() %>%
+    filter(across(c(lon, lat), ~ !is.na(.x))) %>%
+    mutate(Shape = list(st_point(c(lon, lat), dim = "XY"))) %>%
+    st_as_sf() %>% st_set_crs(sites_crs) %>%
+    st_transform(st_crs(reaches_nhd_fields)) %>%
+    st_geometry()
+  message('matching flowlines with reaches...')
+  flowline_indices <- nhdplusTools::get_flowline_index(flines = reaches_nhd_fields,
+                                                       points = sites_sf,
+                                                       max_matches = max_matches,
+                                                       search_radius = search_radius) %>%
```

Comment:
Is 0.1 the default `search_radius` in nhdplusTools? Effectively, this nearest neighbor function will search within that radius (~10 km, I think) and if any segments are returned, we just want the closest one (e.g. maybe that line is 5 m or 5 km away). One thing I hadn't considered before is that with this approach, we may be incorrectly matching samples taken in small, headwater streams to larger river segments, which are the only ones represented by the PRMS lines. Do you think that's true?

I'm not suggesting any changes to the code here, perhaps just something to be aware of (and without looking at the returned matches more closely, I'm not sure whether it's actually an issue that comes up or how often).

Reply:
It looks like 50% of sites have offsets at or below 0.005 (~0.5 km, I think), but some can be higher, up to our max `search_radius`

![offsets](https://user-images.githubusercontent.com/8785034/147144415-80b302b4-a4c3-42ee-94c9-6c32b642aec0.png)

Reply:
I hadn't thought of the headwater streams not being in the network that is being matched to. I think that's good to keep in mind. I'll make an issue for this so we don't forget. For this first model, I was planning on using just the most well-observed sites, so I'm guessing that those will be on a bit bigger streams (ie., I'm guessing that they will be correctly matching to our PRMS segments).

Reply:
OK, sounds good and I would agree with your assessment of the most observed sites. _If_ this is happening, perhaps it's just a matter of adjusting the search radius. As a first impression/ballpark, I would think a GPS point shouldn't plot more than 300-500m away from the intended stream. Thanks for making the new issue.